draft-ietf-jsonpath-base.md

```diff
@@ -1411,82 +1410,101 @@ a type system is first introduced.
-| `OptionalBooleanType`     | `Value(true)`, `Value(false)`, `Nothing` | `BooleanType`                                  |
-| `BooleanType`             | `Value(true)`, `Value(false)`            |                                                |
-| `OptionalNodesType`       | `Nodes(nl)`, `Nothing`                   | `OptionalNodeType`                             |
+a non-empty nodelist.
+
+| Type                 | Abstract Instances                       |
+| :--                  | :----------------                        |
+| `ValueType`          | `Value(v)`, `Nothing`                    |
+| `BooleanType`        | `BooleanTrue`, `BooleanFalse`            |
+| `NodelistType`       | `Nodes(nl)`                              |
 {: #tbl-types title="Function extension type system"}
 
 Notes:
 
-* `OptionalNodeOrValueType` is an abstraction of a `comparable` (which may appear on either side of a comparison or as a function argument).
-* `OptionalNodeType` is an abstraction of a Singular Path.
-* `ValueType` is an abstraction of a primitive value.
-* `BooleanType` is an abstraction of a primitive value that is either
-  `true` or `false`.
-* `OptionalValueType` is an abstraction of a primitive value that may
-  alternatively be absent (`Nothing`).
-* `OptionalNodesType` is an abstraction of a `filter-path` (which appears
+* `ValueType` is an abstraction of a JSON value or `Nothing`.
+* `BooleanType` is an abstraction of a the result of a `test-expr`.
+  Its `BooleanTrue` and `BooleanFalse` values are not relatable to
+  the JSON literals `true` and `false` and have no syntactical representation in JSON Path.
+* `NodelistType` is an abstraction of a `filter-path` (which appears
   in a test expression or as a function argument).
+  Members of `NodelistType` have no syntactical representation in JSON Path.
 
 The abstract instances above can be obtained from the concrete representations in {{tbl-typerep}}.
 
 | Abstract Instance | Concrete Representations                                           |
 | :---------------: | :----------------------:                                           |
-| `Node(n)`         | Singular Path resulting in a nodelist containing just the node `n` |
 | `Value(v)`        | JSON value `v`                                                     |
-| `Nothing`         | Singular Path or `filter-path` resulting in an empty nodelist      |
-| `Nodes(nl)`       | `filter-path` resulting in the non-empty nodelist `nl`             |
+| `Nothing`         | A representation of the absence of a JSON value, distinct from the JSON literal `null` |
+| `Nodes(nl)`       | `filter-path` resulting in the nodelist `nl`, which may or may not be empty  |
 {: #tbl-typerep title="Concrete representations of abstract instances"}
 
-The following subtype relationships depend on coercion:
+### Type Conversions {#type-conv}
+
+The following type conversions may occur:
```

Comment:
Type conversions seem to have to same downsides as type coercions in the current spec. We might be able to get away from conversions/coercions altogether by listing equivalent types where they are used.

Reply:
I listed these as "conversion" because I tend to see "coercion" as a value adjustment within the same time.  For instance, if a value should be constrained betwee 0 and 100, but -5 is submitted, the app could error or _coerce_ the value to 0.

But a conversion exists between values of different types.

I think the conversions are necessary.  The _types_ aren't equivalent, and neither are their values.  For example, you can't go from `BooleanFalse` to an empty nodelist (nor do we have any reason to).  It's a one-way conversion to make nodelists operate properly within expressions.